### ISSUE TITLE
Question: Acceleration in a position controller- How to properly configure a motor for ros_control. 

### ISSUE THREAD

**Issue by facmarques** [ros-controls/ros_control#350]
Hi, we are working on a new robot and we are using ros_control. But we have a couple of questions about how to properly configure a motor in ros_control. So I will put the steps we took here so with your feedback we can produce documentation on how this is supposed to be done (we didn't find any so if it exists sorry :) )

Background information: 
We are using position_controllers/JointTrajectoryController controlling prismatic joints with stepper motors.

### *Step 1 - Motor Configuration**
For what we understood we first started by configuring the motor for the maximum speed and maximum acceleration it can safely reach in the motor's own software suite.
  
### ** Step 2 - Populating the URDF and Joint_limits.yaml**
Then after calculating the speed and acceleration converting it to meters.
We set that information into the URDF (only speed). After we also filled the has_velocity_limits and max velocity in joint_limits.yaml.

Apparently the joint was working correctly but when we checked the error (desired vs actual position) it was very large of about 0.5 meters (3 meter joint). While the error increases in the acceleration phase it stabilizes when the motor reaches the top speed and of course decreases when stopping.

We assume that the problem is that the acceleration is not correcly parameterized. But we can not find where to put it. The acceleration limits on the joint_limits are not enforced by the _joint_limits_interface::PositionJointSoftLimitsInterface_. So how can the position controller know the acceleration limits of our motor?

Of course we can severly decrease the velocity that in turn decreases the error. 
But we think this is not a good approach because the controller is creating a smooth trajectory but it is out of reach of our motor's acceleration. 

So basically the question is how to properly configure the acceleration limits on ros_control?

**Comment by bmagyar** [comment]
Are you using the `joint_trajectory_controller` throught MoveIt? 
The joint_limits.yaml file should allow for specifying acceleration limits like [so](https://github.com/pal-robotics/tiago_moveit_config/blob/indigo-devel/config/joint_limits_steel.yaml#L7) which then *should* be passed to the controller. 

You are probably onto something... 

While `joint_limits_interface` [is able to enforce acceleration limits](https://github.com/ros-controls/ros_control/blob/melodic-devel/joint_limits_interface/include/joint_limits_interface/joint_limits_interface.h#L437) and the [param server parsing](https://github.com/ros-controls/ros_control/blob/melodic-devel/joint_limits_interface/include/joint_limits_interface/joint_limits_rosparam.h#L131) parses acceleration limits, the URDF-parser [seems to skip over parsing acceleration limits](https://github.com/ros-controls/ros_control/blob/melodic-devel/joint_limits_interface/include/joint_limits_interface/joint_limits_urdf.h#L70). If you need this feature, we are happy to review a pull request implementing it, can help guiding along the way if need be.

**Comment by mathias-luedtke** [comment]
> While joint_limits_interface is able to enforce acceleration limits

Only for the velocity interface.

> the URDF-parser seems to skip over parsing acceleration limits.

URDF does store acceleration limits.
They have to get provided by [paramters](https://github.com/ros-controls/ros_control/blob/melodic-devel/joint_limits_interface/include/joint_limits_interface/joint_limits_rosparam.h#L130).

**Comment by facmarques** [comment]
First, thank you for the fast reply as I know that this is not a support forum :). We are producing documentation based on our experience and will provide to you guys for reviewing and if it makes sense to introduce to the wiki.

So today I was going to update the issue with some information we gathered in our tests. But you guys beat me to it.

**_Are you using the joint_trajectory_controller throught MoveIt?_**

Yes we are. But the answer is not that simple and straightforward. 

We are using ros_canopen to control our nanotec motors in prismatic joints. We modified it and will probably suggest a pull request to improve the compatibility with more motors and a homing service and so on. Our part is not yet clean enough for that but we will get there. Also, we need to set the velocity for each position command in position profile and are struggling on how to do it.

So first, before integrating with moveit we were testing the movement only using the rqt_gui. That we later found out creates the trajectory only taking into account the URDF and so only speed like @ipa-mdl says in his post.  So the first trajectory point is already at the maximum speed defined in the URDF (please correct me if I am wrong).

Also, if the hardware interface reads the controller joint_limits.yaml and we have acceleration and velocity limits (in our case with a position controller only velocity) and they are lower (vs URDF) the error is even greater. 

Also in the documentation maybe a clarification is needed that only lower than urdf limits are enforced. While this makes sense and is easy to understand while viewing the code we remember reading that it overrides the limits of the URDF and that made us confused at first.

Now the moveit part :). Yes while we are using moveit and setting the acceleration limits in the moveit_config . We are having trouble on how to account for the jerk compensation. The motor acceleration in the beginning is slower because of the jerk and it never recovers. But this is a moveit question that we need to look further into it.
By the way when we don’t use motors own jerk compensation this happens:

![trapezoidal_bit5 1](https://user-images.githubusercontent.com/20779718/43893406-b9c387ae-9bc6-11e8-8d74-6a37b71b278e.png)

The motor is stopping and starting :(.

Sorry for the long post.

Best Francisco

**Comment by mathias-luedtke** [comment]
> We are using ros_canopen to control our nanotec motors in prismatic joints.

[Not all limit interfaces work properly  with `ros_canopen`](https://github.com/ros-industrial/ros_canopen/issues/141). At least acceleration limits lead to this ripple motion because the limit interface assumes closes-loop control and does not take the delay of 2 cycles into account.

**Comment by lucasw** [comment]
> Also in the documentation maybe a clarification is needed that only lower than urdf limits are enforced.

I updated the wiki to remove the ambiguity:
https://github.com/ros-controls/ros_control/wiki/joint_limits_interface/8554620c9027515ce3c6c461a3a002874b8f28a6

Does that look correct?

(it seems like my update should be reviewed before other people see it?  Haven't used github wikis that much)
